calico/felix/splitter.py

```diff
@@ -66,14 +70,20 @@ def apply_snapshot(self, rules_by_prof_id, tags_by_prof_id,
         # takes longer than this timer to apply then we might do the cleanup
         # before the snapshot is finished.  That would cause dropped packets
         # until applying the snapshot finishes.
-        gevent.spawn_later(self.config.STARTUP_CLEANUP_DELAY,
-                           self.trigger_cleanup)
+        if not self._cleanup_scheduled:
+            _log.info("No cleanup scheduled, scheduling one.")
+            gevent.spawn_later(self.config.STARTUP_CLEANUP_DELAY,
+                               functools.partial(self.trigger_cleanup,
+                                                 async=True))
+            self._cleanup_scheduled = True
 
+    @actor_message()
     def trigger_cleanup(self):
         """
         Called from a separate greenlet, asks the managers to clean up
         unused ipsets and iptables.
         """
+        self._cleanup_scheduled = False
```

Comment:
trigger_cleanup() is now an actor_message so it'll get queued up on the UpdateSplitter's queue, that means it can't run concurrently with apply_snapshot() so this is safe.